chart/config/conditions.yaml

```diff
@@ -0,0 +1,131 @@
+  api:
+    cc_uploader: true
+    cloud_controller_ng: true
+    file_server: true
+    loggr-udp-forwarder: true
+    policy-server-internal: true
+    policy-server: true
+    route_registrar: true
+    statsd_injector: true
+  asactors:
+    operator:      features.autoscaler.enabled
+    scalingengine: features.autoscaler.enabled
+    scheduler:     features.autoscaler.enabled
+  asapi:
+    golangapiserver:  features.autoscaler.enabled
+    metricsforwarder: features.autoscaler.enabled
+    route_registrar:  features.autoscaler.enabled
+  asdatabase:
+    postgres: features.autoscaler.enabled && !features.autoscaler.mysql.enabled
+  asmetrics:
+    eventgenerator: features.autoscaler.enabled
+    metricsserver:  features.autoscaler.enabled
+  asnozzle:
+    metricsgateway: features.autoscaler.enabled
+  auctioneer:
+    auctioneer: !features.eirini.enabled
+  cc-worker:
+    cloud_controller_worker: true
+  credhub:
+    credhub:         features.credhub.enabled
+    route_registrar: features.credhub.enabled
+  diego-api:
+    bbs: true
+    locket: true
+    loggr-udp-forwarder: true
+    silk-controller: true
+  diego-cell:
+    bosh-dns-adapter:    !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    garden:              !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    loggr-udp-forwarder: !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    netmon:              !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    rep:                 !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    route_emitter:       !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    silk-daemon:         !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+    vxlan-policy-agent:  !features.eirini.enabled && !features.multiple_cluster_mode.control_plane.enabled
+  doppler:
+    doppler: true
+  log-api:
+    loggregator_trafficcontroller: true
+    reverse_log_proxy: true
+    reverse_log_proxy_gateway: true
+    route_registrar: true
+  log-cache:
+    log-cache-cf-auth-proxy: true
+    log-cache-gateway: true
+    log-cache: true
+    log-cache-nozzle: true
+    route_registrar: true
+  nats:
+    nats: true
+    nats-tls: true
+  router:
+    gorouter: true
+    loggr-udp-forwarder: true
+  routing-api:
+    routing-api: features.routing_api.enabled
+  scheduler:
+    cc_deployment_updater:        true
+    cloud_controller_clock:       true
+    loggr-syslog-binding-cache:   true
+    loggr-udp-forwarder:          true
+    service-discovery-controller: true
+    ssh_proxy:                    !features.eirini.enabled
+    statsd_injector:              true
+    tps:                          !features.eirini.enabled
+  singleton-blobstore:
+    blobstore:       features.blobstore.provider == singleton
```

Comment:
While technically correct, seeing complex strings with punctuation and whitespace without surrounding quotes always makes me nervous. I would prefer if we always quoted condition expressions.

Reply:
Note that I am not even sure if this kind of condition works with your `_config.condition` macro. There is a question at the top of the file about that.

Reply:
BTW, forgot to add: the condition as written is not currently supported by the `_config.condition` evaluator:

1. There is no comparison operator `==`; the only implemented operations are grouping `()`, negation `!`, and `&&` and or `||`, in the precedence order listed here. So every operand is always treated as a boolean.

2. There is no support for literal strings (because they would all be considered `true`, except for the empty string being `false`. The term `singleton` would currently be evaluated as the value of `$.Values.singleton`, which doesn't exist, so would be considered `true`.

Now, I'm all for extending the `_config.condition` mechanism as needed, but right now we should look how these features are being defined. Normally there should be a `feature.external_database.enabled` property or something like that. This is an issue I want to open, for us to create a more consistent structure for enabling/disabling features, and how to configure them. Once we do that, the current `_config.condition` mechanism may still be sufficient for the current use-case. Revising the feature settings is outside the scope of this PR though.

Reply:
Done with commit 7eb9b9ff3ea2e0cf8f54eab43f2d08fc5d0ee875

Reply:
I agree. The `features.blobstore.provider` currently supports only `s3` and `singleton`, and the difference is external vs local.
The provider tests use either string.

I am all in for switching to a proper boolean `features.external_blobstore.enabled` and then updating the tests using`features.blobstore.provider` to that. I would retain the features.blobstore` hierarchy tough, because that contains the S3 configuration when `provider == s3` i.e. external. And I do not wish to move that around. That said, if you consider movign better, I can do that too.

@viovanov Do you agree to make this switch ? I would do it as part of this PR then.

Reply:
I would not want to modify how `blobstore` is configured as part of a PR that adds memory limits to containers. If we can't make it work without changing the config settings, then maybe this PR needs to be blocked until the other change has been done.

That said, I really don't want to change our settings piecemeal over time; there should be a single release that moves everything around to the proper place, with safeguards in `unsupported` and tight schema restrictions, but afterwards things should be rather stable. This is our external API, after all.

So I think this needs to be thought through comprehensively, and not one feature at a time. I will create a separate issue for this soonish.

About re-arranging blobstore config settings, there is an example of how I thought we should be doing this in https://github.com/cloudfoundry-incubator/quarks-operator/issues/895#issuecomment-616817788 under the `Example` heading.

I'm not sure if I still agree with my previous suggestion, but that is also something that I would want to at least consider before making any changes.

Reply:
@andreas-kupries yes - let's do it
who doesn't like consistency? 😄

Reply:
@jandubois  @viovanov 

> can't make it work without changing the config settings

Is each config file run through regular helm templating as part of its integration ?
Because if yes we could maybe use something like

```
{{- if eq $.Values.features.blobstore.provider "s3" }}
features:
  external_blobstore:
    enabled: true
{{- else }}
features:
  external_blobstore:
    enabled: true
{{- end}}

to generate this new bool key, from a config file named to be very early in the operation.

Alternatively, put this into the config support code, no config file required.

If that works, we can then handle the restructuring of the blobstore setup wholly separate, with only the config code proposed above the overlapping point (the restructure would remove it again).

Reply:
> Is each config file run through regular helm templating as part of its integration ?

No, it is not; it is loaded verbatim, just like `values.yaml`. I think it may be possible to add templating to config files, but I would have to look into it first before I can confirm for sure.

Reply:
Don't bother with templating for config then. ... Should be possible to do that i nthe memory templating to come, with dynamic modification of the `.Values` tree, similar to what you do when loading the config files.